src/charm.py

```diff
@@ -252,6 +333,312 @@ def _on_nginx_route_broken(self, _: Any) -> None:
+            self.unit.status = WaitingStatus("waiting for relation")
+            return
+        for hostname in revoke_list:
+            old_csr = self._tls.get_relation_data_field(
+                f"csr-{hostname}", tls_certificates_relation, self.app
+            )
+            if not old_csr:
+                continue
+            if JujuVersion.from_environ().has_secrets:
+                try:
+                    secret = self.model.get_secret(label=f"private-key-{hostname}")
+                    secret.remove_all_revisions()
+                except SecretNotFoundError:
+                    LOGGER.warning("Juju secret for %s already does not exist", hostname)
+                    continue
+            try:
+                self._tls.pop_relation_data_fields(
+                    [f"key-{hostname}", f"password-{hostname}"],
+                    tls_certificates_relation,
+                    self.app,
+                )
+                peer_relation = self.model.get_relation("nginx-peers")
+                if peer_relation:
+                    self._tls.pop_relation_data_fields(
+                        [f"key-{hostname}", f"password-{hostname}"], peer_relation, self.app
+                    )
+            except KeyError:
+                LOGGER.warning("Relation data for %s already does not exist", hostname)
+            self.certificates.request_certificate_revocation(
+                certificate_signing_request=old_csr.encode()
+            )
+
+    def _on_certificate_invalidated(self, event: CertificateInvalidatedEvent) -> None:
+        """Handle the TLS Certificate invalidation event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        tls_certificates_relation = self.model.get_relation("certificates")
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for peer relation to be created")
+            event.defer()
+            return
+        if event.reason == "revoked":
+            hostname = self._tls.get_hostname_from_csr(
+                tls_certificates_relation, self.app, event.certificate_signing_request
+            )
+            self._certificate_revoked([hostname])
+        if event.reason == "expired":
+            self._on_certificate_expiring(event)
+        self.unit.status = MaintenanceStatus("Waiting for new certificate")
+
+    def _on_all_certificates_invalidated(self, _: AllCertificatesInvalidatedEvent) -> None:
+        """Handle the TLS Certificate relation broken event.
+
+        Args:
+            _: The event that fires this method.
+        """
+        hostnames = self.get_additional_hostnames()
+        for hostname in hostnames:
+            try:
+                secret = self.model.get_secret(label=f"private-key-{hostname}")
+                secret.remove_all_revisions()
+            except SecretNotFoundError:
+                LOGGER.warning("Juju secret for %s already does not exist", hostname)
```

Comment:
Shouldn't we manage the case when juju doesn't support secrets ?

Reply:
Not really. the purpose of this method is to erase any remnants of the tls relation. The relation data will be deleted from the certificates relation whether we use secrets or not because the relation is extinct. If we do have secrets, however, we want to erase the existing secrets, completing the erase of the relevant data.

Reply:
Hmm I need to dig again, but I think I've read that even when a relation is removed and we get the event, there could be remnant of the data in the databag which could require explicitly flushing these data